TombEngine/Objects/TR4/Trap/tr4_chain.cpp

```diff
@@ -1,41 +1,296 @@
 	void ControlChain(short itemNumber)
 	{
 		auto& item = g_Level.Items[itemNumber];
 
-		if (item.TriggerFlags)
+		if (!TriggerActive(&item))
+			return;
+
+		if (item.TriggerFlags < 0)
 		{
-			item.ItemFlags[2] = 1;
-			item.ItemFlags[3] = 75;
+			int damage = abs(item.TriggerFlags);
+
+			item.ItemFlags[PendulumFlags::COLLIDE_BITS] = 1;
+			item.ItemFlags[PendulumFlags::FLAME_EFFECT] = 1; // Set the item on fire when collide.
 
 			if (TriggerActive(&item))
 			{
-				*(int*)&item.ItemFlags[0] = 0x787E;
+				item.ItemFlags[PendulumFlags::PUSH_ITEM] = 1;
 				AnimateItem(&item);
+
+				auto flameMesh = item.ItemFlags[PendulumFlags::FLAME_MESH];
+
+				auto pos = GetJointPosition(item, flameMesh, Vector3i(0, 260, 0));
+				auto angle = GetBoneOrientation(item, 5);
+
+				unsigned char r = item.ItemFlags[PendulumFlags::FIRE_COLOR_R];
+				unsigned char g = item.ItemFlags[PendulumFlags::FIRE_COLOR_G];
+				unsigned char b = item.ItemFlags[PendulumFlags::FIRE_COLOR_B];
+
+				Vector3 flameColor1 = Vector3::Zero;
+				Vector3 flameColor2 = Vector3::Zero;
+
+				if (item.ItemFlags[PendulumFlags::FIRE_COLOR_R] == 0 &&
+					item.ItemFlags[PendulumFlags::FIRE_COLOR_G] == 0 &&
+					item.ItemFlags[PendulumFlags::FIRE_COLOR_B] == 0)
+				{
+					 r = 51 - ((GetRandomControl() / 16) & 6);
+					 g = 44 - ((GetRandomControl() / 64) & 6);
+					 b = GetRandomControl() & 10;	
+				}
+				else
+				{
+					auto sourceColorR = std::clamp(r + 0.2f, 0.0f, 1.0f);
+					auto sourceColorG = std::clamp(g + 0.2f, 0.0f, 1.0f);
+					auto sourceColorB = std::clamp(b + 0.2f, 0.0f, 1.0f);
+
+					flameColor1 = Vector3(sourceColorR, sourceColorG, sourceColorB);
+					flameColor2 = Vector3(r, g, b);
+				}
+
+				SpawnDynamicLight(pos.x, pos.y, pos.z, 12, r, g, b);
+
+				r += 125 - ((GetRandomControl() / 16) & 4);
+				g += 98 - ((GetRandomControl() / 16) & 8);
+
+				auto color = Color(r / (float)CHAR_MAX, g / (float)CHAR_MAX, b / (float)CHAR_MAX);
+
+				SpawnDynamicFogBulb(pos.ToVector3(), PENDULUM_FIRE_FOG_RADIUS, PENDULUM_FIRE_FOG_DENSITY, color);
+				TriggerPendulumFlame(itemNumber, pos, color);
+				TriggerPendulumSpark(pos, angle, PENDULUM_FLAME_SPARK_LENGHT, 1, color);
+				TriggerFireFlame(pos.x, pos.y, pos.z, FlameType::Trail, flameColor1, flameColor2);
+
 				return;
 			}
 		}
-		else
+		else if (item.TriggerFlags > 0)
 		{
-			item.ItemFlags[3] = 25;
+			item.ItemFlags[PendulumFlags::COLLIDE_BITS] = 1;
+			item.ItemFlags[PendulumFlags::FLAME_EFFECT] = 0;
 
 			if (TriggerActive(&item))
 			{
-				*(int*)&item.ItemFlags[0] = 0x780;
+				item.ItemFlags[PendulumFlags::PUSH_ITEM] = 1;
+				AnimateItem(&item);
+				return;
+			}
+		} 
+		else
+		{
+			if (TriggerActive(&item))
+			{
+				*(int*)&item.ItemFlags[PendulumFlags::PUSH_ITEM] = 0x780;
 				AnimateItem(&item);
 				return;
 			}
 		}
 
-		*(int*)&item.ItemFlags[0] = 0;
+		item.ItemFlags[PendulumFlags::PUSH_ITEM] = 0;
+	}
+
+	void CollideChain(short itemNumber, ItemInfo* playerItem, CollisionInfo* coll)
```

Comment:
Where this collision function code originate from? Before that, chain was using generic sphere-box collision.

Reply:
changed to Test harmJoints